```{r}
#| label: load-packages
#| include: false

library(tidyverse)
library(palmerpenguins)
```

## A. Installing the required software

1. Install MAFFT.
<!-- Insert logos of applications to be downloaded -->
  - For Windows, follow the steps outlined [here](https://mafft.cbrc.jp/alignment/software/ubuntu_on_windows.html).
  - For Mac, this may be done using Homebrew. Open Terminal and run the command: `brew install mafft`.

2. Visit [MEGA’s official website](https://www.megasoftware.net/) and download the appropriate version of MEGA 11 for your machine.

3. Download BEAST 2 version 2.7.7 from [BEAST 2's official website](https://www.beast2.org).

4. Install Tracer v1.7.2 by downloading the appropriate file for your machine [here](https://github.com/beast-dev/tracer/releases/tag/v1.7.2). 

5. Install FigTree v1.4.4 [here](https://github.com/rambaut/figtree/releases). As in Step 4, download the appropriate file for your machine: dmg for Mac, zip for Windows, and tgz for Linux.

## B. Download SARS-CoV-2 metadata and sequences from GISAID
1. Proceed to [https://gisaid.org/](https://gisaid.org/).
2. Under EpiCov, click Search and select at least two viruses. #insert screenshot
3. Download the corresponding FASTA file.


## C. Quality Check with Nextclade
1. Proceed to [clades.nextstrain.org](clades.nextstrain.org).
2. Import your FASTA file under “Add sequence data” and click “Run.”
3. After analysis, examine the six QC metrics for each sequence.
4. Remove or flag sequences that fail quality metrics.
5. Save the filtered dataset as sars-cov-2_QC_passed.fasta.

## D. Alignment using MAFFT
1. Type Ubuntu on your search bar.
2. Click Ubuntu 18.04.5/Ubuntu 20.04.5. (Note: For Mac, open Terminal.)
3. Navigate to the directory containing your data file (sars-cov-2.fasta) using the cd command as follows:
`cd win-home/Desktop/Phylogenetic\ Analysis\Training/Data/`

4. To know the file name, type `ls`.

5. Then type `mafft`.

6. Highlight the file name in step 4 (sars-cov-2.fasta) then press Ctrl+Shift+C (Ctrl+C for Mac) then right click in the input file section and paste the file name.

7. For the output file, type `sars-cov-2_align.fasta`.

8. For the output format, type `4`.

9. For strategy, choose auto by typing `1`. (Note: You may also choose your desired strategy.)

10. For additional arguments, just press `Enter`.
11. If you are okay with your inputs, press `Enter`.
12. Wait for the program to finish the alignment. When you see - - More - - (0%), it means that the alignment is finished. (Note: For Mac, the last line should be the name of the output file.)
13. Exit the application.

The output file (sars-cov-2_align.fasta) will be generated in the same directory.

## E. Trimming using MEGA 11

1. Open MEGA 11.
2. Click `File` then `Open a File/Session`. Navigate to your aligned fasta file generated in Part D.
3. Choose `Align`.
4. To remove leading gaps in sequences, locate the last column containing gaps (-). Hold Shift and click the first column with gaps to select the entire region, then press Delete (fn+Delete for Mac) to remove these columns.
5. To remove trailing gaps, navigate to the end of the sequences. Hold Shift and click to select all columns with gaps from the first gap to the end of the alignment, then press Delete (fn+Delete for Mac) to remove these columns.
6. You now have the same length for all the sequences. Now, you need to search for gaps in between the sequences by using Ctrl+F then press “-” (minus sign) then Enter.
7. Inspect all columns with gaps. You may use the Find Next function under Search to find the next gap. If the column has more than 2 nucleotides present, DO NOT DELETE the column. Otherwise, delete it.
8. After trimming everything, delete the reference fasta file.
9. Go to Data then Export Alignment then FASTA format. Save the file in the training folder, then rename it to sars-cov-2_align_trim.fasta. Then press Enter.
10. Exit the application.

## F. Checking Temporal Signal
1. Launch TempEst and select the tree file from the previous section, then click Open.
2. Navigate to the Parse Dates tab and select "Define just by its order." Choose "last" from the Order dropdown menu. If your data only includes years, select "Parse as a Number." For other configuration of dates in the taxon labels, you may refer to [this documentation](https://beast.community/tip_dates). 
3. Proceed to the Root-to-tip tab. On the left panel, choose the most appropriate procedure under Function, and tick the box for Best-fitting root.
4. Take note of the value for Slope (rate).

For more information, see [this documentation ](https://beast.community/tempest_tutorial).

## G. Generating and running the XML file
1. In your Desktop, open the BEAST folder, then open BEAUti.
2. Click `File`, `Import Alignment`, and select the trimmed file.
3. In the data type of alignment, select `nucleotide` then `OK`.
4. Go to the `Tip Dates` tab and select `Use Tip Dates`. To do this, the sequences used for the analysis should have been labeled with tip dates. Select the appropriate format of the dates in the `as dates with format`.
5. Click `Auto-configure`. Click `use everything` then select `after last` then type the character that was used in the name of your sequences before the dates. Then click `OK`.
6. Go to the `Site Model` tab and select the substitution model to be used. For uniformity, select `HKY` model then set the gamma category count to `4`.
7. Go to the `Clock Model` tab and select the molecular clock to be used. For uniformity, select `Strict Clock`. Use the clock rate obtained in Section F.
8. Go to the `Priors` tab and select the Tree Prior to be used. For uniformity, select `Phylodynamics: Birth Death SIR (serial)`. You can leave the other parameters as default. 
9. Go to the `MCMC` tab and specify the length of the chain and the frequency of sampling the posterior distribution. It is recommended to have a minimum of 100,000 trees sampled from the posterior. For analysis involving estimation of multiple model parameters or evolutionary dynamics, the length of the chain will need to be increased accordingly. For this training, set the chain length to 50,000. Set the `tracelog`, `screenlog` and `treelog` to 5,000. 
10. Click `Save` in the `File` tab then type `sars-cov-2.xml`. Make sure that that file type is xml.
11. Open BEAST. Click `Choose File` and navigate to the directory where the xml file is located. Select the file and click `Open` to load it. Tick the box for `Use BEAGLE library if available`.
12. Click `Run` to start the tree estimation process. You will see a java window displaying the progress of the analysis. The analysis will automatically stop once the length of the chain has been reached. BEAST will output log files containing the samples obtained from the posterior for all parameters being estimated and a trees file containing all the phylogenetic trees sampled from the posterior distribution.

## H. Diagnosing convergence problems using Tracer
1. Open the Tracer software. Click `File` and select `Import Trace File`. Navigate to the directory where the log file (sars-cov-2.log) is located. Select the log file and click `Open` to load it.
2. Examine the Effective Sampling Size (ESS) of each continuous parameter. Ideally, all important tree parameters under investigation should have a **minimum ESS of 200**. You can utilize the different options available to visualize the posterior distribution of each estimated parameter.
3. Trace statistics can be improved by increasing the MCMC length of the chain. Additionally, it is possible to simultaneously run the same xml file and combine the output files to improve the ESS values of each parameter.
4. The LogCombiner software can be used to combine one or more log files (or tree files) given that all have been analyzed using the same parameters.

## I. Annotating the phylogenetic trees
1. Open the **Tree Annotator** software. Specify the burn in as number of states, typically 10% of the MCMC length of chain.
2. Click Choose File in `Input Tree File`. Navigate to the directory where the trees file (sars-cov-2-sars-cov-2_align_trim.trees) generated by BEAST is located. Select the trees file and click `Open` to load it.
3. Click Choose File in `Output File`. Input the filename `sars-cov-2_annotate.mcc` to be used for the Maximum Clade Credibility (MCC) Tree. Click `Save`.
4. Click `Run` and wait for the annotation to finish.
5. Open the **FigTree software**. Click `File` and select `Open`. Navigate to the directory where the MCC Tree is located. Select the tree file and click `Open` to load it.
6. You can annotate the MCC tree by selecting your preferred annotations that can be found at the left side of the panel. The resulting MCC tree can also be exported into an image or a pdf file.